### DISCUSSION TITLE
How to render task.fields values in progress bar

### DISCUSSION THREAD

**Discussion by EdLeafe** [Textualize/rich#1114] [Q&A] [answered]
My use case requires that more than one text value be displayed with the progress bar for a task. For example, a part number as well as the description of the part. I can add a `part_num=1234` when creating the task, but I haven't been able to figure out how to display that. If I use the entire dict, it can display that, but I can't reference a specific value.

Example:
```
import random
import time

from rich.progress import BarColumn, Progress, TextColumn


def run_progress(progress):
    tot = 20
    curr = 0
    with progress:
        task = progress.add_task("0123ABC", total=tot, part_name="Gear")
        while curr < tot:
            tm = random.randrange(1, 5)
            curr += tm
            time.sleep(tm / 10.0)
            progress.update(task, advance=tm)


progress = Progress(
    "[progress.description]{task.description}",
    TextColumn(text_format="Name: {task.fields}"),
    BarColumn(),
    "[magenta]{task.completed} of {task.total}",
)
run_progress(progress)
```

This script shows the following output:

`0123ABC Name: {'part_name': 'Gear'}  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20 of 20`

Of course, I don't want to display the `fields` dict; I just want the value for `part_name`. But if I change the `TextColumn` to this:

`TextColumn(text_format="Name: {task.fields['part_name']}")`

... I get the following error:

```
Traceback (most recent call last):
  File "/Users/ed/projects/ngc-cli/progress.py", line 24, in run_progress
    task = progress.add_task("Processing...", total=tot, part_name="Gear")
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 978, in add_task
    self.refresh()
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 899, in refresh
    self._live_render.set_renderable(self.get_renderable())
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 905, in get_renderable
    renderable = RenderGroup(*self.get_renderables())
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 910, in get_renderables
    table = self.make_tasks_table(self.tasks)
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 935, in make_tasks_table
    widget = column(task)
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 184, in __call__
    renderable = self.render(task)
  File "/Users/ed/venvs/ngc/lib/python3.9/site-packages/rich/progress.py", line 274, in render
    _text = self.text_format.format(task=task)
KeyError: "'part_name'"
```

So what is the trick to displaying the values in `task.fields`?

**Comment by willmcgugan** [comment, marked as answer]
Progress uses `str.format` there which has a weird syntax for looking up a key. If you remove the single quotes around part_name it should work.

You could also extend a `TextColumn` if you need more that a format string.

**Reply by EdLeafe** [reply]
Guess I never ran into the weird syntax before. Removing the quotes works, but man, does it look odd!

Thanks!

**Comment by jfthuong** [comment]
@willmcgugan  Would that not be interesting to highlight it in the documentation?

I know you have:

`You can add them to a format string with the following syntax: "extra info: {task.fields[extra]}".`

But to most of Pythonista, they would assume you would need quotes... we could put something like ("note that quotes shall be omitted because `rich` is using `str.format`).
